lib/internal/async/fs.js

```diff
@@ -0,0 +1,947 @@
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'path',
+                                 ['string', 'Buffer', 'URL']);
+    }
+
+    const req = new FSReqWrap();
+    req.oncomplete = promise;
+    _mkdir(_makeLong(path), mode, req);
+  } catch (err) {
+    promiseReject(promise, err);
+  }
+  return promise;
+}
+
+function mkdtemp(prefix, options) {
+  const promise = createPromise();
+  try {
+    options = getOptions(options, {});
+    if (!prefix || typeof prefix !== 'string') {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE',
+                                 'prefix',
+                                 'string');
+    }
+    nullCheck(prefix);
+
+    const req = new FSReqWrap();
+    req.oncomplete = promise;
+    _mkdtemp(`${prefix}XXXXXX`, options.encoding, req);
+  } catch (err) {
+    promiseReject(promise, err);
+  }
+  return promise;
+}
+
+function readdir(path, options) {
+  const promise = createPromise();
+  try {
+    options = getOptions(options, {});
+    handleError(path = getPathFromURL(path));
+    nullCheck(path);
+
+    if (typeof path !== 'string' && !Buffer.isBuffer(path)) {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'path',
+                                 ['string', 'Buffer', 'URL']);
+    }
+
+    const req = new FSReqWrap();
+    req.oncomplete = promise;
+    _readdir(_makeLong(path), options.encoding, req);
+  } catch (err) {
+    promiseReject(promise, err);
+  }
+  return promise;
+}
+
+function readlink(path, options) {
+  const promise = createPromise();
+  try {
+    options = getOptions(options, {});
+    handleError(path = getPathFromURL(path));
+    nullCheck(path);
+
+    if (typeof path !== 'string' && !Buffer.isBuffer(path)) {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'path',
+                                 ['string', 'Buffer', 'URL']);
+    }
+
+    const req = new FSReqWrap();
+    req.oncomplete = promise;
+    _readlink(_makeLong(path), options.encoding, req);
+  } catch (err) {
+    promiseReject(promise, err);
+  }
+  return promise;
+}
+
+function rename(oldPath, newPath) {
+  const promise = createPromise();
+  try {
+    handleError(oldPath = getPathFromURL(oldPath));
+    handleError(newPath = getPathFromURL(newPath));
+
+    nullCheck(oldPath);
+    nullCheck(newPath);
+
+    if (typeof oldPath !== 'string' && !Buffer.isBuffer(oldPath)) {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'oldPath',
+                                 ['string', 'Buffer', 'URL']);
+    }
+
+    if (typeof newPath !== 'string' && !Buffer.isBuffer(newPath)) {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'newPath',
+                                 ['string', 'Buffer', 'URL']);
+    }
+
+    const req = new FSReqWrap();
+    req.oncomplete = promise;
+
+    _rename(_makeLong(oldPath),
+            _makeLong(newPath),
+            req);
+  } catch (err) {
+    promiseReject(promise, err);
+  }
```

Comment:
Have you tried making all of these functions `async function`s to have implicit try/catch+reject? I think unless you’re having any `await` in it that’s all there they are.

Reply:
Yep, was considering that already :-)